README
					LNAMP 一键安装包 - Readme

LNAMP 一键安装包是什么?
——————————————————————

LNAMP 一键安装包是一个用 Linux Shell 编写的可以为 CentOS/RadHat/Fedora、Debian/Ubuntu/Raspbian VPS(VDS) 或独立主机安装 LNMP(Nginx/MySQL/PHP)、LAMP(Apache/MySQL/PHP)、LNAMP(Nginx/Apache/MySQL/PHP) 生产环境的 Shell 程序。同时提供一些实用的辅助工具如：虚拟主机管理、FTP用户管理、Nginx、MySQL/MariaDB、PHP的升级、常用缓存组件的安装、重置 MySQL root 密码、502自动重启、日志切割、SSH 防护 DenyHosts/Fail2Ban、备份等许多实用脚本。

--------

LNAMP修改版作者:  shines77 <shines77@cloudbuses.com>
LNAMP修改版官网： http://lnamp.cloudbuses.com

--------

LNMP原版作者:  licess <admin@lnmp.org>
LNMP原版官网： http://lnmp.org

--------------------------------------------------------------------------------

安装
—————
详细安装教程参考：http://lnamp.cloudbuses.com/install.html

安装前建议使用 screen，执行：screen -S lnamp 后,

执行：wget -c http://mirror.cloudbuses.com/lnamp/lnamp-1.0-full.tar.gz && tar zxf lnamp-1.0-full.tar.gz && cd lnamp-1.0-full && ./install.sh {lnmp|lamp|lnamp}

如断线可使用 screen -r lnamp 恢复。

常用功能
—————————

以下操作需在 lnamp-1.0-full 目录下执行：

FTP服务器：执行：./pureftpd.sh 安装，http://yourIP/ftp/ 进行管理，也可使用 lnamp ftp {add|list|del} 进行管理。

升级脚本：

执行：./upgrade.sh 按提示进行选择
也可以直接使用参数：./upgrade.sh {nginx|mysql|mariadb|php|phpa|m2m}

参数: nginx   可升级至任意 Nginx 版本。
参数: mysql   可升级至任意 MySQL 版本，MySQL 升级风险较大，虽然会自动备份数据，依然建议自行再备份一下。
参数: mariadb 可升级已安装的 Mariadb，虽然会自动备份数据，依然建议自行再备份一下。
参数: m2m     可从 MySQL 升级至 Mariadb，虽然会自动备份数据，依然建议自行再备份一下。
参数: php     仅适用于LNMP，可升级至大部分 PHP 版本。
参数: phpa    可升级 LAMP/LNAMP 的 PHP 至大部分版本。


缓存加速：

执行: ./addons.sh {install|uninstall} {eaccelerator|xcache|memcached|opcache|redis|imagemagick|ioncube}

参数: xcache       安装时需选择版本和设置密码， http://yourIP/xcache/ 进行管理，用户名 admin，密码为安装 xcache 时设置的。
参数: redis
参数: memcached    可选择 php-memcache 或 php-memcached 扩展。
参数: opcache      http://yourIP/ocp.php 进行管理。
参数: eaccelerator 安装 eaccelerator。

请勿安装多个缓存类扩展模块，多个可能导致网站出现问题。

图像处理：

参数: imageMagick imageMagick 路径：/usr/local/imagemagick/bin/。

解密：
可选1，执行：./ionCube.sh 安装。

其他：
可选1，执行：./php5.2.17.sh 可安装一个不与 LNMP 冲突的 PHP 5.2.17 单独存在，目录在 /usr/local/php52/，使用时需要将 nginx 虚拟主机配置文件里的 php-cgi.sock 修改为 php-cgi52.sock 即可调用 PHP 5.2.17。
可选2，执行：./reset_mysql_root_password.sh 可重置 MySQL/MariaDB 的 root 密码。
可选3，执行：./check502.sh  可检测 php-fpm 是否挂掉,502报错时重启，配合 crontab 使用。
可选4，执行：./cut_nginx_logs.sh 日志切割脚本。
可选5，执行：./remove_disable_function.sh 运行此脚本可删掉禁用函数。
可选6，如需卸载 LNMP、LAMP 或 LNAMP，可执行：./uninstall.sh 按提示选择即可卸载。

状态管理
—————————

LNMP/LMAP/LNAMP状态管理：lnmp {start|stop|reload|restart|kill|status}
Nginx状态管理：          lnmp nginx 或 /etc/init.d/nginx {start|stop|reload|restart}
MySQL状态管理：          lnmp mysql 或 /etc/init.d/mysql {start|stop|restart|reload|force-reload|status}
MariaDB状态管理：        lnmp mariadb 或 /etc/init.d/mariadb {start|stop|restart|reload|force-reload|status}
PHP-FPM状态管理：        lnmp php-fpm 或 /etc/init.d/php-fpm {start|stop|quit|restart|reload|logrotate}
PureFTPd状态管理：       lnmp pureftpd 或 /etc/init.d/pureftpd {start|stop|restart|kill|status}
Apache状态管理：         lnmp httpd 或 /etc/init.d/httpd {start|stop|restart|graceful|graceful-stop|configtest|status}

虚拟主机管理
————————————
添加：lnmp vhost add
删除：lnmp vhost del
列出：lnmp vhost list

相关图形界面
————————————
PHPMyAdmin：          http://yourIP/phpmyadmin/
phpinfo：             http://yourIP/phpinfo.php
PHP探针：              http://yourIP/p.php
PureFtp用户管理：      http://yourIP/ftp/
Xcache管理界面：       http://yourIP/xcache/
Zend Opcache管理界面： http://yourIP/ocp.php

LNMP相关目录文件
————————————————

目录位置

Nginx：/usr/local/nginx/
MySQL：/usr/local/mysql/
MariaDB：/usr/local/mariadb/
PHP：/usr/local/php/
PHPMyAdmin：/home/wwwroot/default/phpmyadmin/
默认虚拟主机网站目录：/home/wwwroot/default/
Nginx日志目录：/home/wwwlogs/

配置文件：

Nginx主配置文件：/usr/local/nginx/conf/nginx.conf
MySQL/MariaDB配置文件：/etc/my.cnf
PHP配置文件：/usr/local/php/etc/php.ini
PureFtpd配置文件：/usr/local/pureftpd/pure-ftpd.conf
PureFtpd MySQL配置文件：/usr/local/pureftpd/pureftpd-mysql.conf
Apache配置文件：/usr/local/apache/conf/httpd.conf


技术支持
————————

技术支持论坛：http://bbs.vpser.net/forum-25-1.html


ChangeLog.txt
更新记录

2015年6月1日发布LNMP 1.2

. 完全重写优化安装脚本、升级脚本及组件安装等大部分脚本；
. 状态管理、虚拟主机管理、FTP管理及数据库管理全部采用lnmp命令替换，去掉/root/vhost.sh和/root/lnmp脚本；
. Nginx、MySQL/MariaDB、PHP的升级脚本整合为upgrade.sh;
. 缓存、加速、imageMagick、ionCube等的安装卸载采用addons.sh进行；
. 增加SSH防护工具DenyHosts、Failban安装脚本及DenyHosts被封IP删除脚本；
. 增加新的备份脚本backup.sh；
. 支持的Linux发行版：CentOS 5-7、RHEL 6-7、Fedora 20-21、Aliyun、Debian 6-8、Raspbian、Ubuntu 10.10-15.04；
. 支持ARM平台，Raspberry Pi已测试；
. MySQL安装时可选5.1.73、5.5.42、5.6.23，MariaDB可选5.5.42、10.0.17；
. PHP安装时可选5.2.17、5.3.29、5.4.41、5.5.25、5.6.9；
. 安装时增加内存优化jemalloc或gperftools；
. Apache安装时可选2.2.29或2.4.12；
. Apache日志统一到/home/wwwlogs/下；
. 更新Nginx至1.8.0；
. 更新Zend Guard Loader至支持PHP 5.6；
. Nginx配置文件进行了调整，启用php和pathinfo采用include方式，增加ssl范例配置；


2014年6月1日发布LNMP 1.1

. 调整优化脚本安装脚本；
. 安装时增加MariaDB选项；
. 增加MySQL升级脚本 upgrade_mysql.sh；
. 增加MySQL升级到MariaDB脚本 upgrade_mysql2mariadb.sh；
. 增加MariaDB升级脚本 upgrade_mariadb.sh；
. 增加Redis安装脚本 redis.sh；
. 增加OPcache安装脚本 opcache.sh；
. 增加MySQL/MariaDB root密码重置脚本；
. 添加pathinfo支持，去掉include pathinfo的#，在try_files前加#重启nginx即可使用；
. 重写nginx init.d脚本；
. 优化PHP升级脚本；
. 优化LNMP升级到LNMPA脚本，PHP可选PHP5.2或5.3,调整编译参数和配置文件；
. 优化LNMPA PHP升级脚本，支持5.4,5.5；
. 优化Xcache安装脚本，增加xcache管理界面；
. 优化Memcached安装脚本；
. 优化其他相关脚本使其支持MySQL的同时可支持MariaDB；
. 优化Nginx部分配置；
. 优化Pureftpd安装脚本；
. 优化Nginx配置文件，增加部分参数，去掉注释即可使用；
. 优化PHP-FPM参数，增加slowlog慢日志选项，可修改request_slowlog_timeout参数启用；
. 更新wordpress伪静态；
. 更新Nginx状态页为http://IP/nginx_status；
. 更新Nginx至1.6.0；
. 更新MySQL至5.1.73,5.5.37；
. 更新PHP至5.3.28；
. 更新Xcache至3.1.0；
. 更新ImageMagick至6.8.8-9；
. 更新phpMyAdmin至4.2.2；
. 去除fcgi.conf，使用nginx自带的fastcgi.conf；
. 去除Nginx日志多余配置，采用统一日志名称；
. 去除PHP编译参数--with-curlwrappers；


2013年6月1日发布LNMP 1.0

. 调整优化安装脚本，自动记录安装日志至/root/lnmp-install.log安装失败时可以提供此文件；
. 安装时增加PHP 5.3和MySQL 5.5版本选择，可以自由选择；
. 删除安装时需要输入域名的选择，以免对新生造成误导；
. 默认站点目录更换到/home/wwwroot/default/ ；
. 增加proftpd安装脚本 proftpd.sh和proftpd用户创建脚本proftpd_vhost.sh；
. 增加xcache安装脚本；
. 增加502重启脚本；
. 增加php 5.2.17安装脚本，适合已升级5.3，而有程序需要php 5.2的用户；
. 添加nginx日志切割脚本 cut_nginx_logs.sh;
. 添加php禁用函数删除脚本remove_disable_function.sh;
. 将memecache扩展安装脚本移至memcached.sh；
. 优化eaccelerator，增加测试版本选择，可以使其支持php 5.4；
. 优化ionCube Loader安装脚本，增加对php 5.4的支持；
. PHP禁用函数里去掉fsockopen和pfsockopen；
. 增加shopex、dedecms、drupal、ecshop伪静态文件；
. PHP 5.3版本增加request_terminate_timeout php脚本运行时间限制参数，如果有采集程序将该值修改为0；
. nginx配置增加IE6禁用gzip；
. nginx版本升级为1.2.7；
. 增加支持PHP 5.4的最新版ZendGuardLoader；


2012年4月23日发布LNMP 0.9

. 优化安装过程及错误处理；
. 修正eaccelerator和ionCube的PHP版本判断错误的问题；
. 修正memcached pid创建失败的问题
. 修正pdo_mysql安装方法；
. PHP增加部分禁用函数；
. 优化Nginx升级脚本；
. 去除vsftpd安装脚本；
. 修正访问不存在的PHP时返回404；
. 更新Nginx版本；
. 更新PHP探针；
. 增加phpwin伪静态
. 升级过程增加停止LNMP相关服务；
. 修正Nginx log_format问题；

2011年4月27日发布LNMP 0.7

. 优化安装过程，修复多处安装错误处理；
. 添加LNMP至LNMPA升级包；
. 重写PureFTP安装包，实现只需要输入密码即可安装成功；
. Nginx升级至0.8.54、MySQL升级至5.1.54、PHP升级至5.2.17；
. 优化PHP升级脚本；
. 优化Nginx升级脚本；
. 添加LNMPA PHP版本升级工具；
. 优化eAccelerator自动安装脚本，支持LNMPA同时多次安装都不会出现多次加载；
. 优化ionCube自动安装脚本，支持LNMPA同时多次安装都不会出现多次加载；；
. 去除了Suhosin；

2011年2月8日发布LNMP 0.6

. 优化安装过程，修复多处安装错误处理；
. 添加ubuntu专用安装脚本ubuntu.sh；
. 修善pureftpd安装脚本，增加自动开机启动；

2011年2月15日发布Nginx版本升级工具

2011年3月7日发布Nginx日志切割工具

2011年3月15日发布PHP自动升级工具

使用新探针，修复邮件测试功能

2010年10月9日发布LNMP 0.5

. 升级Nginx到0.7.67、MySQL到5.1.48、PHP升级到5.2.14、PHPMyAdmin升级到3.3.7
. 为安全将Nginx日志移动到/home/wwwlogs/下面；
. 更改32位64位判别方式；
. 增加系统时间调整，防止时间错误造成编译出错；
. 优化eAccelerator安装程序；
. 增加ionCube安装程序；
. 虚拟主机添加可以直接运行：/root/vhost.sh；
. 修正/root/lnmp状态判断错误。

2010年11月27日 更新lnmp文件上传大小限制为50MB
2010年12月7日 修正/root/pureftpd状态判断错误。
2010年12月10日 更新Nginx支持IPv6
2011年1月17日 将PHP执行时间有30秒调整为300秒。
2011年1月17日 安装脚本添加自动关闭SELINUX

2010年3月28日发布LNMP 0.4

. 添加LNMP状态管理
. 添加PureFTPd状态管理
. PHP增加suhosin
. 增加VsFTPd、PureFTPd安装及管理
. 增强Nginx虚拟主机添加功能，添加多域名、伪静态支持、日志是否启用的功能
. 升级Nginx到0.7.65、MySQL到5.1.44

2009年11月15日发布LNMP 0.3

. 添加虚拟主机添加程序

2009年3月2日发布LNMP 0.2

2009年1月19日发布LNMP0.1


include/main.sh
#!/bin/bash

# which Memory Allocator do you want to install?

MemeoryAllocator_Selection()
{
    SelectMalloc="1"
    echo ""
    Echo_Yellow "You have 3 options for your Memory Allocator install:"
    echo ""
    echo "1: Don't install Memory Allocator. (Default)"
    echo "2: Install Jemalloc"
    echo "3: Install TCMalloc"
    echo ""
    read -p "Enter your choice (1, 2 or 3): " SelectMalloc

    echo ""
    case "${SelectMalloc}" in
        1)
            Echo_Blue "You will install not install Memory Allocator."
        ;;
        2)
            Echo_Blue "You will install JeMalloc."
        ;;
        3)
            Echo_Blue "You will Install TCMalloc."
        ;;
        *)
            Echo_Blue "No input, You will not install Memory Allocator."
            SelectMalloc="1"
        ;;
    esac
    echo ""

    if [ "${SelectMalloc}" = "1" ]; then
        MySQL51MAOpt=''
        MySQL55MAOpt=''
        MariaDBMAOpt=''
        NginxMAOpt=''
    elif [ "${SelectMalloc}"="2" ]; then
        MySQL51MAOpt='--with-mysqld-ldflags=-ljemalloc'
        MySQL55MAOpt="-DCMAKE_EXE_LINKER_FLAGS='-ljemalloc' -DWITH_SAFEMALLOC=OFF"
        MariaDBMAOpt=''
        NginxMAOpt="--with-ld-opt='-ljemalloc'"
    elif [ "${SelectMalloc}"="3" ]; then
        MySQL51MAOpt='--with-mysqld-ldflags=-ltcmalloc'
        MySQL55MAOpt="-DCMAKE_EXE_LINKER_FLAGS='-ltcmalloc' -DWITH_SAFEMALLOC=OFF"
        MariaDBMAOpt="-DCMAKE_EXE_LINKER_FLAGS='-ltcmalloc' -DWITH_SAFEMALLOC=OFF"
        NginxMAOpt='--with-google_perftools_module'
    fi
}

# which MySQL Version do you want to install?

MySQL_Selection()
{
    DBSelect="2"
    echo ""
    Echo_Yellow "You have 5 options for your DataBase install:"
    echo ""
    echo "1: Install MySQL 5.1.73"
    echo "2: Install MySQL 5.5.42 (Default)"
    echo "3: Install MySQL 5.6.23"
    echo "4: Install MariaDB 5.5.42"
    echo "5: Install MariaDB 10.0.17"
    echo ""
    read -p "Enter your choice (1, 2, 3, 4 or 5): " DBSelect

    echo ""
    case "${DBSelect}" in
        1)
            Echo_Blue "You will install MySQL 5.1.73."
        ;;
        2)
            Echo_Blue "You will install MySQL 5.5.42."
        ;;
        3)
            Echo_Blue "You will Install MySQL 5.6.23."
        ;;
        4)
            Echo_Blue "You will install MariaDB 5.5.42."
        ;;
        5)
            Echo_Blue "You will install MariaDB 10.0.17."
        ;;
        *)
            Echo_Blue "No input, You will install MySQL 5.5.42 (Default)."
            DBSelect="2"
        ;;
    esac
    echo ""

    if [ "${DBSelect}"="4" ] || [ "${DBSelect}"="5" ]; then
        MySQL_Bin="/usr/local/mariadb/bin/mysql"
        MySQL_Config="/usr/local/mariadb/bin/mysql_config"
        MySQL_Dir="/usr/local/mariadb"
    else
        MySQL_Bin="/usr/local/mysql/bin/mysql"
        MySQL_Config="/usr/local/mysql/bin/mysql_config"
        MySQL_Dir="/usr/local/mysql"
    fi
}

# do you want to enable or disable the InnoDB Storage Engine?

InnoDB_StorageEngine_Selection()
{
    InstallInnodb="y"
    echo ""
    Echo_Yellow "Do you want to enable or disable the InnoDB Storage Engine?"
    echo ""
    read -p "Default enable, Enter your choice [y/N]: " InstallInnodb

    echo ""
    case "${InstallInnodb}" in
        [yY][eE][sS]|[yY])
            Echo_Blue "You will enable the InnoDB Storage Engine."
        ;;
        [nN][oO]|[nN])
            Echo_Blue "You will disable the InnoDB Storage Engine!"
        ;;
        *)
            Echo_Blue "No input, The InnoDB Storage Engine will enable."
            InstallInnodb="y"
        ;;
    esac
    echo ""
}

# Input MySQL root password

Input_MySQL_RootPWD()
{
    read -p "Please enter the password: " MysqlRootPWD
    if [ "${MysqlRootPWD}" = "" ]; then
        echo ""
        Echo_Magenta "You have no input, Mysql root password will be use the default value."
        echo ""
        read -p "Are you sure use the default password: '${MysqlRootDefaultPWD}' ? [Y/n]: " MysqlRootUseDefaultPWD

        echo ""
        case "${MysqlRootUseDefaultPWD}" in
            [yY][eE][sS]|[yY])
                Echo_Cyan "You agree to use the default Mysql root password '${MysqlRootDefaultPWD}'."
                MysqlRootUseDefaultPWD='y'
                MysqlRootPWD="${MysqlRootDefaultPWD}"
                MysqlRootConfirmPWD="${MysqlRootDefaultPWD}"
            ;;
            [nN][oO]|[nN])
                Echo_Red "Info: You do not agree to use the default Mysql root password, please try again."
                MysqlRootUseDefaultPWD='n'
                echo ""
                Input_MySQL_RootPWD
            ;;
            *)
#                Echo_Cyan "No input, You agree to use the default Mysql root password '${MysqlRootDefaultPWD}'."
#                MysqlRootUseDefaultPWD="y"
#                MysqlRootPWD="${MysqlRootDefaultPWD}"
#                MysqlRootConfirmPWD="${MysqlRootDefaultPWD}"
                Echo_Cyan "No input, You do not agree to use the default Mysql root password, please try again."
                MysqlRootUseDefaultPWD='n'
                echo ""
                Input_MySQL_RootPWD
            ;;
        esac
    else
        read -p "     Confirm the password: " MysqlRootConfirmPWD
        if [ "${MysqlRootConfirmPWD}" = "" ]; then
            MysqlRootConfirmPWD="${MysqlRootDefaultPWD}"
        fi
        if [ "${MysqlRootPWD}" != "${MysqlRootConfirmPWD}" ]; then
            echo ""
            Echo_Red "Error: two time passwords are not equal, please try again."
            echo ""
            Input_MySQL_RootPWD
        fi
    fi
}

# Set MySQL root password

MySQL_RootPWD_Setting()
{
    MysqlRootDefaultPWD="mysql2015"
    MysqlRootPWD=""
    MysqlRootConfirmPWD=""
    Echo_Yellow "Please setup the root password of MySQL. (Default password is: ${MysqlRootDefaultPWD})"
    echo ""

    Input_MySQL_RootPWD
    echo ""
    Echo_Blue "Your MySQL root password is: ${MysqlRootPWD}"
    echo ""
}

# Which PHP Version do you want to install?

PHP_Selection()
{
    PHPSelect="3"
    echo ""
    Echo_Yellow "You have 5 options for your PHP install:"
    echo ""
    echo "1: Install PHP 5.2.17"
    echo "2: Install PHP 5.3.29"
    echo "3: Install PHP 5.4.41"
    echo "4: Install PHP 5.5.25 (Default)"
    echo "5: Install PHP 5.6.9"
    echo ""
    read -p "Enter your choice (1, 2, 3, 4 or 5): " PHPSelect

    echo ""
    case "${PHPSelect}" in
        1)
            Echo_Blue "You will install PHP 5.2.17."
        ;;
        2)
            Echo_Blue "You will install PHP 5.3.29."
        ;;
        3)
            Echo_Blue "You will Install PHP 5.4.41."
        ;;
        4)
            Echo_Blue "You will install PHP 5.5.25."
        ;;
        5)
            Echo_Blue "You will install PHP 5.6.9."
        ;;
        *)
            Echo_Blue "No input, You will install PHP 5.5.25 (Default)."
            PHPSelect="4"
        ;;
    esac
    echo ""
}

# Which Nginx Version do you want to install?

Nginx_Selection()
{
    NginxSelect="3"
    echo ""
    Echo_Yellow "You have 3 options for your Nginx install:"
    echo ""
    echo "1: Install Nginx 1.4.7"
    echo "2: Install Nginx 1.6.3"
    echo "3: Install Nginx 1.8.0 (Default)"
    echo ""
    read -p "Enter your choice (1, 2 or 3): " NginxSelect

    echo ""
    case "${NginxSelect}" in
        1)
            Echo_Blue "You will install Nginx 1.4.7."
        ;;
        2)
            Echo_Blue "You will install Nginx 1.6.3."
        ;;
        3)
            Echo_Blue "You will Install Nginx 1.8.0."
        ;;
        *)
            Echo_Blue "No input, You will install Nginx 1.8.0 (Default)."
            NginxSelect="3"
        ;;
    esac
    echo ""
}

# Which Apache Version do you want to install?

Apache_Selection()
{
    ApacheSelect="1"
    echo ""
    Echo_Yellow "You have 2 options for your Apache install:"
    echo ""
    echo "1: Install Apache 2.2.29 (Default)"
    echo "2: Install Apache 2.4.10"
    echo ""
    read -p "Enter your choice (1 or 2): " ApacheSelect

    echo ""
    if [ "${ApacheSelect}" = "1" ]; then
        Echo_Blue "You will install Apache 2.2.29."
    elif [ "${ApacheSelect}" = "2" ]; then
        Echo_Blue "You will install Apache 2.4.10."
    else
        Echo_Blue "No input, You will install Apache 2.2.29 (Default)."
        ApacheSelect="1"
    fi
    echo ""
}

# Set Server Administrator Email Address

Admin_Email_Setting()
{
    ServerAdmin=""
    echo ""
    read -p "Please enter Administrator Email Address: " ServerAdmin
    echo ""
    if [ "${ServerAdmin}" == "" ]; then
        Echo_Blue "Server Administrator Email will set to webmaster@example.com !"
        ServerAdmin="webmaster@example.com"
    else
        echo "========================================================"
        Echo_Blue "Server Administrator Email is: ${ServerAdmin}"
        echo "========================================================"
    fi
    echo ""
}

# Get the selections about install

Display_Selection()
{
    echo "=========================================================="

    MemeoryAllocator_Selection

    echo "=========================================================="

    MySQL_Selection

    echo "=========================================================="

    InnoDB_StorageEngine_Selection

    MySQL_RootPWD_Setting

    echo "=========================================================="

    PHP_Selection

    echo "=========================================================="

    Nginx_Selection

    if [[ "${Stack}" = "lamp" || "${Stack}" = "lnamp" || "${Stack}" = "test" ]]; then

        echo "=========================================================="

        Apache_Selection

        echo "=========================================================="

        Admin_Email_Setting
    fi
}

Press_Install()
{
    echo ""
    echo "Press any key to install or Press Ctrl + C to cancel ..."
    OLDCONFIG=`stty -g`
    stty -icanon -echo min 1 time 0
    dd count=1 2>/dev/null
    stty ${OLDCONFIG}
    . include/version.sh
}

Press_Start()
{
    echo ""
    echo "Press any key to start or Press Ctrl + C to cancel ..."
    OLDCONFIG=`stty -g`
    stty -icanon -echo min 1 time 0
    dd count=1 2>/dev/null
    stty ${OLDCONFIG}
}

Print_Sys_Info()
{
    cat /etc/issue
    cat /etc/*-release
    uname -a
    MemTotal=`free -m | grep Mem | awk '{print  $2}'`
    echo "Memory is: ${MemTotal} MB "
    df -h
}

Check_OS_Is_64Bit()
{
    if [[ `getconf WORD_BIT` = '32' && `getconf LONG_BIT` = '64' ]]; then
        Is_64bit='y'
    else
        Is_64bit='n'
    fi
}

Check_Is_ARM()
{
    if uname -m | grep -Eqi "arm"; then
        Is_ARM='y'
    fi
}

Install_LSB()
{
    if [ "$PM" = "yum" ]; then
        yum -y install redhat-lsb
    elif [ "$PM" = "apt" ]; then
        apt-get update
        apt-get install -y lsb-release
    fi
}

Get_Linux_Dist_Version()
{
    Install_LSB
    eval ${DISTRO}_Version=`lsb_release -rs`
    eval echo "${DISTRO} \${${DISTRO}_Version}"
}

Get_Linux_Dist_Name()
{
    if grep -Eqi "CentOS" /etc/issue || grep -Eq "CentOS" /etc/*-release; then
        DISTRO='CentOS'
        PM='yum'
    elif grep -Eqi "Red Hat Enterprise Linux Server" /etc/issue || grep -Eq "Red Hat Enterprise Linux Server" /etc/*-release; then
        DISTRO='RHEL'
        PM='yum'
    elif grep -Eqi "Aliyun" /etc/issue || grep -Eq "Aliyun" /etc/*-release; then
        DISTRO='Aliyun'
        PM='yum'
    elif grep -Eqi "Fedora" /etc/issue || grep -Eq "Fedora" /etc/*-release; then
        DISTRO='Fedora'
        PM='yum'
    elif grep -Eqi "Debian" /etc/issue || grep -Eq "Debian" /etc/*-release; then
        DISTRO='Debian'
        PM='apt'
    elif grep -Eqi "Ubuntu" /etc/issue || grep -Eq "Ubuntu" /etc/*-release; then
        DISTRO='Ubuntu'
        PM='apt'
    elif grep -Eqi "Raspbian" /etc/issue || grep -Eq "Raspbian" /etc/*-release; then
        DISTRO='Raspbian'
        PM='apt'
    else
        DISTRO='unknow'
    fi
    Check_OS_Is_64Bit
}

Get_RHEL_Version()
{
    Get_Linux_Dist_Name
    if [ "${DISTRO}" = "RHEL" ]; then
        if grep -Eqi "release 5." /etc/redhat-release; then
            echo "Current Version: RHEL Ver 5"
            RHEL_Ver='5'
        elif grep -Eqi "release 6." /etc/redhat-release; then
            echo "Current Version: RHEL Ver 6"
            RHEL_Ver='6'
        elif grep -Eqi "release 7." /etc/redhat-release; then
            echo "Current Version: RHEL Ver 7"
            RHEL_Ver='7'
        fi
    fi
}

Download_Files()
{
    local URL=$1
    local FileName=$2
    if [ -s "${FileName}" ]; then
        echo "${FileName} [found]"
    else
        echo "Error: ${FileName} not found!!! download now ..."
        wget -c ${URL}
    fi
}

Tar_Cd()
{
    local FileName=$1
    local DirName=$2
    cd ${cur_dir}/src
    [[ -d "${DirName}" ]] && rm -rf ${DirName}
    echo "Uncompress ${FileName} ..."
    tar zxf ${FileName}
    echo "cd ${DirName} ..."
    cd ${DirName}
}

StartUp()
{
    init_name=$1
    echo "Add ${init_name} service at system startup ..."
    if [ "$PM" = "yum" ]; then
        chkconfig --add ${init_name}
        chkconfig ${init_name} on
    elif [ "$PM" = "apt" ]; then
        update-rc.d -f ${init_name} defaults
    fi
}

Remove_StartUp()
{
    init_name=$1
    echo "Removing ${init_name} service at system startup ..."
    if [ "$PM" = "yum" ]; then
        chkconfig ${init_name} off
        chkconfig --del ${init_name}
    elif [ "$PM" = "apt" ]; then
        update-rc.d -f ${init_name} remove
    fi
}

Get_PHP_Ext_Dir()
{
    Cur_PHP_Version=`/usr/local/php/bin/php -r 'echo PHP_VERSION;'`
    if echo "${Cur_PHP_Version}" | grep -Eqi '^5.2.'; then
       zend_ext_dir="/usr/local/php/lib/php/extensions/no-debug-non-zts-20060613/"
    elif echo "${Cur_PHP_Version}" | grep -Eqi '^5.3.'; then
       zend_ext_dir="/usr/local/php/lib/php/extensions/no-debug-non-zts-20090626/"
    elif echo "${Cur_PHP_Version}" | grep -Eqi '^5.4.'; then
       zend_ext_dir="/usr/local/php/lib/php/extensions/no-debug-non-zts-20100525/"
    elif echo "${Cur_PHP_Version}" | grep -Eqi '^5.5.'; then
       zend_ext_dir="/usr/local/php/lib/php/extensions/no-debug-non-zts-20121212/"
    elif echo "${Cur_PHP_Version}" | grep -Eqi '^5.6.'; then
       zend_ext_dir="/usr/local/php/lib/php/extensions/no-debug-non-zts-20131226/"
    fi
}

Check_Stack()
{
    if [[ -s /usr/local/nginx/sbin/nginx && -s /usr/local/apache/bin/httpd && -s /usr/local/apache/conf/httpd.conf && -s /etc/init.d/httpd && ! -s /usr/local/php/sbin/php-fpm ]]; then
        Get_Stack="lnamp"
    elif [[ -s /usr/local/php/bin/php-cgi || -s /usr/local/php/sbin/php-fpm ]] && [[ -s /usr/local/php/etc/php-fpm.conf && -s /etc/init.d/php-fpm && -s /usr/local/nginx/sbin/nginx ]]; then
        Get_Stack="lnmp"
    elif [[ -s /usr/local/apache/bin/httpd && -s /usr/local/apache/conf/httpd.conf && -s /etc/init.d/httpd && ! -s /usr/local/php/sbin/php-fpm ]]; then
        Get_Stack="lamp"
    else
        Get_Stack="unknow"
    fi
}

Check_DB()
{
    if [[ -s /usr/local/mariadb/bin/mysql && -s /usr/local/mariadb/bin/mysqld_safe && -s /etc/my.cnf ]]; then
        MySQL_Bin="/usr/local/mariadb/bin/mysql"
        MySQL_Config="/usr/local/mariadb/bin/mysql_config"
        MySQL_Dir="/usr/local/mariadb"
        Is_MySQL="n"
        DB_Name="mariadb"
    else
        MySQL_Bin="/usr/local/mysql/bin/mysql"
        MySQL_Config="/usr/local/mysql/bin/mysql_config"
        MySQL_Dir="/usr/local/mysql"
        Is_MySQL="y"
        DB_Name="mysql"
    fi
}

Verify_DB_Password()
{
    Check_DB
    read -p "verify your current database root password: " DB_Root_Password
    ${MySQL_Bin} -uroot -p${DB_Root_Password} -e "quit"
    if [ $? -eq 0 ]; then
        echo "MySQL root password correct."
    else
        echo "MySQL root password incorrect! Please check!"
        Verify_DB_Password
    fi
    if [ "${DB_Root_Password}" = "" ]; then
        Verify_DB_Password
    fi
}
